clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManager.java

```diff
@@ -137,16 +137,23 @@ public void onHeartbeatSuccess(ShareGroupHeartbeatResponse response) {
                     "already leaving the group.", memberId, memberEpoch);
             return;
         }
-        if (state == MemberState.UNSUBSCRIBED && maybeCompleteLeaveInProgress()) {
+        if (state == MemberState.UNSUBSCRIBED && responseData.memberEpoch() < 0 && maybeCompleteLeaveInProgress()) {
             log.debug("Member {} with epoch {} received a successful response to the heartbeat " +
-                    "to leave the group and completed the leave operation. ", memberId, memberEpoch);
+                "to leave the group and completed the leave operation. ", memberId, memberEpoch);
```

Comment:
nit: The indentation of this line is now inconsistent with the rest of the method.